### ISSUE TITLE
Improving performance

### ISSUE THREAD

**Issue by MrMerkin** [verlok/vanilla-lazyload#538]
So I've been playing with this for a few days now and have improved some of the performance issues:

- Setting images to opacity 0 on exit with CSS `transition: opacity .1s` (works pretty good but not perfect)
- Drastically reduced CLI by using `SVG` placeholders  in `src` with same aspect ratio, in `base64`
- Trimming down CSS file to reduce render blocking and preloading what I can
- Hosting some scripts locally instead of CDN works better

Working on `srcset` now, but I noticed some specific cases that cause major lag and thought you may know of some workarounds. This is mainly when you scroll and images go out of view before loaded, fade all at once during a scroll event, or scrolling before images fade in fully again after re-entering view causing more loads of new images. If the page is static (no scrolling), typically simultaneous fade in animations don't seem to cause lag and actually looks quite nice—desktop works best, slower machines may still show fps dip.

- **Fast scrolling of the page** (causes worse lag at the first load, less lag after that, but still looks bad and causes fps spikes)
- **jQuery** `.animate scrollTop` (scrolling to top of page or to `#`) if too many images load/fade between start and endpoints
- **Simulating a slower connection** with browser dev tools, images out of view still load and bottleneck
- **Scrolling on a slow/cheap/old computer on a fast internet connection** (tested on my old laptop)
- **Using a mobile device or older computer with high-dpi screen**—load/fade in multiple larger images, so increases lag
- **Too many images in view at once**, bottleneck from multiple simultaneous fade in.

How do you handle these problems so that lag and skipping isn't introduced? Is there  a way to say "if scrolling speed is > than X, then pause the loading, OR check if the image is still in view after Y ms?" Or "if X images are in the load queue, stagger the adding of `loaded` classes so the fade in doesn't overlap?" etc, etc. I know CSS transitions have a delay, but I think that's for multiple animations on a specific id or class, and I don't think it will sequence one animation across multiple elements so when A is finished B will start.

I tried implementing an `.animate scrollTo` with a variable speed based on distance, but either the browser or jQuery does not behave properly doing this (maybe the code I found is bad), and sometimes going to a `#` further down the page is 10 times faster for no reason. Using multiplication or division with speed didn't matter and produce unreliable results. Was either too fast, too slow, but mainly unpredictable speed. Slower motion was better at reducing lag for lazyLoaded images that fade in, but still introduce too much skipping/lag and is jarring.

```javascript
// tried division
$(".offset-scroll").click(function(e) {
		e.preventDefault();
		var dist = jQuery(window).scrollTop();
		var speed = dist / 1.5;
		var loc = $(this).attr("href");
		$('html,body').animate({scrollTop: $(loc).offset().top-100},speed, 'linear');
});
// and multiplication
$(".offset-scroll").click(function(e) {
		e.preventDefault();
		var dist = jQuery(window).scrollTop();
		var speed = dist * 1.5;
		var loc = $(this).attr("href");
		$('html,body').animate({scrollTop: $(loc).offset().top-100},speed, 'linear');
});
```
I also tested `iframe`s that load banner ads. This works really well (for the most part). The problem arises when the ad script fetches a new ad image into the existing iframe—since the iframe has already loaded previously, the new ad image just "pops" in during scroll and causes a massive fps spike. Not sure if there is a way to deal with that.

Also, I couldn't figure a way to make this work with "Native" ads that load asynchronously. I contacted the ad company and they refused to provide `iframe` versions and said manipulating that code was not allowed. Go figure 🙄.

I crush the images I'm using with ImageOptim, so even slower computers/connections have multiple images load at the same time and so they fade in at the same time. Most images range from 10KB to maybe 200KB+ (for large backgrounds @2x). Oddly enough, the smaller images seem to cause the most lag, and not the large background images.

Firefox, Chrome and Brave show the most lag on macOS. Safari seems to handle more images without noticeable lag (though it has other problems with flexbox and such).

At any rate, if you can provide any methods you know of for creating a better user experience in some of these cases I mentioned, I would appreciate it. Pausing loading and staggering fades conditionally might be the way to go. Again, I'm not a programmer by trade, so I may be missing the obvious best practices for this stuff. My guess is it may involve some conditional javascript logic that is far beyond my skill, but idk.

_p.s. sorry for the novel. just trying to be thorough._

**Comment by MrMerkin** [comment]
Also, I noticed today, when I stepped away from the browser for a while and came back, and scrolled up a page, the images no longer faded back in. I assume this is some type of timeout. Let me know if there's a way to avoid this happening or reengage the listener.

Thanks

**Comment by verlok** [comment]
Thank you @MrMerkin for taking the time to write all this question. 

> How do you handle these problems so that lag and skipping isn't introduced? Is there a way to say "if scrolling speed is > than X, then pause the loading, OR check if the image is still in view after Y ms?" Or "if X images are in the load queue, stagger the adding of loaded classes so the fade in doesn't overlap?" etc, etc. 

Unfortunately, the answer is no to all of the above questions. 

A couple of considerations though: 

- providing the code you are using in an isolated environment such as codepen or codesandbox would be really helpful for anyone to help
- how many DOM elements are we talking about? If they are too many, have you considered unloading the images (e.g. restoring the initial placeholder) to free up some browser memory  
- what are you using jQuery for? Meaning, what else is happening in the page?
- you know you could animate and fade using CSS and smooth scroll using CSS, with greater performance

I think you need consultancy on performance improvement beyond the features of this LazyLoad script. Such consultancy goes beyond the basic advice and the level of support I intend to provide here for free. 

I have years of experience in optimising websites for best performance and I would be glad to help. It you feel like hiring me, feel free to reach out. You can find my contacts here https://www.andreaverlicchi.eu/contact/

Thanks

**Comment by MrMerkin** [comment]
I'm using jQuery for Fancybox to display large images in a lightbox and to open a mobile version of the site navigation. Tonight I managed to find a vanilla js alternative for the scroll to top at least, and it doesn't seem to force the browser to reflow. This has made the scrolling performance 1000% better than jQuery, though manual can cause problems. I may be able to run jquery slim with fancybox now that the initial scrolling problem is solved. In a perfect world, I'd dump fancybox and have a custom lightbox, because it's very simple. I just don't have time to figure it out myself.

Not sure if I have any time to put things into codepen, as I can't show the work itself. I'll see if I can manage though.

As far as elements... Header Image, "New" large featured image strip (1600px x 300px), 4 images displaying additional recent items (image sizes vary from small to medium) with text and info, 10 50x50 jpg thumbnails (maybe 6-10kB each) that show other items recently posted. A header strip with page title has small tiling jpg (maybe 10kB), 2 full-width asides with background images in page footer (maybe 120kB each), and a few fonts and text. Using CSS variables for most of the colors. Gutted wordpress of unnecessary bloat. Everything has been crushed as much as possible, preloading what I can, and using CDNs for libraries like jQuery. I'm even conditionally enqueuing the fonts and scripts so there's nothing _not_ being used on each page. Working on srcset now for responsive image sizes.

I am fading with CSS as per our initial conversation, but the overlapping fades tax the CPU if too many images come into view if manually fast scroll the page. The ad swaps also add to this.

No worries, man. Trust, me. If I could afford to hire a developer I wouldn't have taught myself 4 languages and suffered through any of this for the past year and half. I'm a professional artist and a musician. I'd rather not be coding. Pandemic et al have just thrown life into a hell storm and I can't find work. I have to get this project launched asap to try and save my ass, so figured it couldn't hurt to ask. ;) Your script has saved me a lot of time and headaches already and improved performance from the default. Worst case, I'll just make everything static once they first load, or just have certain elements fade in and out and call it a day for the initial site launch. I decided to mark it as a "beta" version anyway. I will keep you in mind for the non-beta version though. I planned to hire someone to consolidate the PHP functions, fix some admin additions, harden the server more than I can manage, setup CSP, and make everything run smoothly.

Thanks for all the help 👍🏻
